receiver/k8sobjectsreceiver/config.go

```diff
@@ -0,0 +1,146 @@
+// limitations under the License.
+
+package k8sobjectsreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sobjectsreceiver"
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/config"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8sconfig"
+)
+
+type mode string
+
+const (
+	PullMode  mode = "pull"
+	WatchMode mode = "watch"
+
+	defaultPullInterval time.Duration = time.Hour
+	defaultMode         mode          = PullMode
+)
+
+var modeMap = map[mode]bool{
+	PullMode:  true,
+	WatchMode: true,
+}
+
+type K8sObjectsConfig struct {
+	Name          string        `mapstructure:"name"`
+	Namespaces    []string      `mapstructure:"namespaces"`
+	Mode          mode          `mapstructure:"mode"`
+	LabelSelector string        `mapstructure:"label_selector"`
+	FieldSelector string        `mapstructure:"field_selector"`
+	Interval      time.Duration `mapstructure:"interval"`
+	gvr           *schema.GroupVersionResource
+}
+
+type Config struct {
+	config.ReceiverSettings `mapstructure:",squash"`
+	k8sconfig.APIConfig     `mapstructure:",squash"`
+
+	Objects []*K8sObjectsConfig `mapstructure:"objects"`
+
+	// For mocking purposes only.
+	makeDiscoveryClient func() (discovery.ServerResourcesInterface, error)
+	makeDynamicClient   func() (dynamic.Interface, error)
+}
+
+func (c *Config) Validate() error {
+
+	validObjects, err := c.getValidObjects()
+	if err != nil {
+		return err
+	}
+	for _, object := range c.Objects {
+		gvr, ok := validObjects[object.Name]
+		if !ok {
+			availableResource := make([]string, len(validObjects))
+			for k := range validObjects {
+				availableResource = append(availableResource, k)
+			}
+			return fmt.Errorf("resource %v not found. Valid resources are: %v", object.Name, availableResource)
+		}
+
+		if object.Mode == "" {
+			object.Mode = defaultMode
+		} else if _, ok := modeMap[object.Mode]; !ok {
+			return fmt.Errorf("invalid mode: %v", object.Mode)
+		}
+
+		if object.Mode == PullMode && object.Interval == 0 {
+			object.Interval = defaultPullInterval
+		}
+
+		object.gvr = gvr
+	}
+	return c.ReceiverSettings.Validate()
+}
+
+func (c *Config) getDiscoveryClient() (discovery.ServerResourcesInterface, error) {
+	if c.makeDiscoveryClient != nil {
+		return c.makeDiscoveryClient()
+	}
+
+	client, err := k8sconfig.MakeClient(c.APIConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.Discovery(), nil
+}
+
+func (c *Config) getDynamicClient() (dynamic.Interface, error) {
+	if c.makeDynamicClient != nil {
+		return c.makeDynamicClient()
+	}
+
+	return k8sconfig.MakeDynamicClient(c.APIConfig)
+}
+
+func (c *Config) getValidObjects() (map[string]*schema.GroupVersionResource, error) {
+	dc, err := c.getDiscoveryClient()
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := dc.ServerPreferredResources()
```

Comment:
Why preferred resources? I probably doesn't return a list of all supported resource, does it?

Reply:
As a k8s user, I can run something in my cluster with a supported but not preferred version. I want to fetch those objects as well.

Reply:
ServerPreferredResources returns all resources with the preferred Version (i.e. latest version).
For example, `horizontalpodautoscalers` is in `autoscaling/v1` and `autoscaling/v2`. Here, `autoscaling/v2` is preferred by the server. So, ServerPreferredResources will have `horizontalpodautoscalers` with `autoscaling/v2` group. 

In my testing, Kubernetes always created a server-preferred version even if I provided an old group version in the manifest.
For example, when I applied the following config:
```yaml
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  name: php-apache
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: php-apache
  minReplicas: 1
  maxReplicas: 10
  targetCPUUtilizationPercentage: 50
```
It created  hpa with apiVersion=autoscaling/v2
```
$ kubectl get hpa php-apache -o yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"autoscaling/v1","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"name":"php-apache","namespace":"default"},"spec":{"maxReplicas":10,"minReplicas":1,"scaleTargetRef":{"apiVersion":"apps/v1","kind":"Deployment","name":"php-apache"},"targetCPUUtilizationPercentage":50}}
  creationTimestamp: "2022-10-13T12:37:35Z"
  name: php-apache
  namespace: default
  resourceVersion: "40370"
  uid: 16c1282e-b708-408c-8b2c-30a28982dd62
...
```

It was also detected the receiver.

Reply:
Ok sounds good. Thanks for explaining!